pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -0,0 +1,482 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.operator.blocks.ProjectionBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.spi.data.FieldSpec;
+
+
+/**
+ * The <code>CaseTransformFunction</code> class implements the CASE-WHEN-THEN-ELSE transformation.
+ *
+ * The SQL Syntax is:
+ *    CASE
+ *        WHEN condition1 THEN result1
+ *        WHEN condition2 THEN result2
+ *        WHEN conditionN THEN resultN
+ *        ELSE result
+ *    END;
+ *
+ * Usage:
+ *    case(${WHEN_STATEMENT_1}, ..., ${WHEN_STATEMENT_N},
+ *         ${THEN_EXPRESSION_1}, ..., ${THEN_EXPRESSION_N},
+ *         ${ELSE_EXPRESSION})
+ *
+ * There are 2 * N + 1 arguments:
+ *    <code>WHEN_STATEMENT_$i</code> is a <code>BinaryOperatorTransformFunction</code> represents <code>condition$i</code>
+ *    <code>THEN_EXPRESSION_$i</code> is a <code>TransformFunction</code> represents <code>result$i</code>
+ *    <code>ELSE_EXPRESSION</code> is a <code>TransformFunction</code> represents <code>result</code>
+ *
+ */
+public class CaseTransformFunction extends BaseTransformFunction {
+  public static final String FUNCTION_NAME = "case";
+  private final List<TransformFunction> _whenStatements = new ArrayList<>();
+  private final List<TransformFunction> _elseThenStatements = new ArrayList<>();
```

Comment:
So the size of this array is _`whenStatements.size() + 1` ? If so, can we assert for this?

Reply:
yes, will add.
There is N WHEN statements and N THEN statements and 1 ELSE statement.